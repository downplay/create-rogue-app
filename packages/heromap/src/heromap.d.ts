declare const lexer: any;
export { lexer as Lexer };
export declare const ParserRules: (
  | {
      name: string;
      symbols: any[];
      postprocess: (d: any) => any;
    }
  | {
      name: string;
      symbols: (
        | string
        | {
            match: RegExp;
            lineBreaks: boolean;
          }
        | {
            type: string;
          }
      )[];
      postprocess?: undefined;
    }
)[];
export declare const ParserStart: string;
