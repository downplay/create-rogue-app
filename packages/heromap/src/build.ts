import { stringify } from "flatted";
import { RNG } from "herotext";
import { createRng } from "../../herotext/src/rng";
import { VirtualGrid, virtualGrid } from "./virtualGrid";
import { BrushNode, BrushNodes } from "./types";
import {
  MapNode,
  NumberNode,
  OperationNode,
  OrOpsNode,
  BrushOpNode,
} from "./types";

type BuildContext = {
  map: MapNode;
  rows: String[][];
  grid: VirtualGrid<MapElement>;
  scope: Record<string, any>;
  rng: RNG;
};

type GlyphTarget = {
  x: number;
  y: number;
  glyph: string;
  index: number;
};

let applyBrushToGlyphTargets: (
  targets: GlyphTarget[],
  brush: BrushNodes | BrushNode,
  context: BuildContext
) => void;

const distributeTargetsToNodes = (
  targets: GlyphTarget[],
  brushes: BrushNode[],
  context: BuildContext
) => {
  // Aggregate nodes into groups
  let totalReserve = 0;
  let totalFraction = 0;
  let fractional = [];
  let unQuantified = [];
  const brushesToTargets = [];

  for (const node of brushes) {
    brushesToTargets.push({
      brush: node.brush,
      targets: [] as GlyphTarget[],
    });
    if (!node.quantifier) {
      unQuantified.push(node);
    } else {
      switch (node.quantifier.type) {
        case "Heromap::PercentageValue": {
          let weight = node.quantifier.value;
          totalFraction += weight;
          fractional.push({
            node,
            weight,
          });
          break;
        }
        case "Heromap::FractionValue":
          let weight = node.quantifier.numerator / node.quantifier.denominator;
          totalFraction += weight;
          fractional.push({
            node,
            weight,
          });
          break;
        default:
          throw new Error("Unhandled quantifier type: " + node.quantifier.type);
      }
    }
  }
  const remainingFraction = 1 - totalFraction;
  const fractionPerUnQuantified =
    remainingFraction > 0 && unQuantified.length > 0
      ? remainingFraction / unQuantified.length
      : 0;
  // Adjust fractions down if we went over 100%.
  const fractionAdjust =
    remainingFraction < 0 && totalFraction > 0 ? 1 / totalFraction : 1;
  // Add the possibility of selecting nothing, if there is remaining fraction to distribute
  const nothing =
    remainingFraction > 0 && fractionPerUnQuantified === 0
      ? [{ weight: remainingFraction, node: null }]
      : [];
  // TODO: Randomly select reserved, leave the rest
  const weightedList = [
    ...fractional.map(({ weight, node }) => ({
      weight: weight * fractionAdjust,
      node,
    })),
    ...unQuantified.map((node) => ({ weight: fractionPerUnQuantified, node })),
    ...nothing,
  ];
  for (const target of targets) {
    const picked =
      weightedList.length === 1
        ? weightedList[0]
        : context.rng.pick(weightedList, "weight");
    if (picked.node) {
      const found = brushesToTargets.find(
        ({ brush }) => brush === picked.node.brush
      );
      found?.targets.push(target);
    }
  }

  for (const brushTarget of brushesToTargets) {
    applyBrushToGlyphTargets(brushTarget.targets, brushTarget.brush, context);
  }
};

applyBrushToGlyphTargets = (
  targets: GlyphTarget[],
  brush: BrushNodes | BrushNode,
  context: BuildContext
) => {
  if (typeof brush === "undefined" || !brush.type) {
    throw new Error("Not a brush node: " + JSON.stringify(brush));
  }
  switch (brush.type) {
    case "Heromap::BrushNode":
      // if (brush.quantifier) {
      // const quantifiedTargets = quantifyTargets(targets, quantifier, rng);
      // }
      applyBrushToGlyphTargets(targets, brush.brush, context);
      break;
    case "Heromap::AndBrushesNode":
      for (const and of brush.brushes) {
        applyBrushToGlyphTargets(targets, and, context);
      }
      break;
    case "Heromap::OrBrushesNode":
      distributeTargetsToNodes(targets, brush.brushes, context);
      break;
    case "Heromap::GlyphNode":
      for (const target of targets) {
        context.rows[target.y][target.x] = brush.glyph;
      }
      break;
    case "Heromap::WordNode":
      if (brush.path.length > 1) {
        throw new Error("Path chaining not supported: " + brush.path.join("."));
      }
      for (const target of targets) {
        context.grid.add(target.x, target.y, {
          type: brush.path[0],
          params: {}, // TODO: Params
        });
      }
      break;

    default:
      throw new Error("Unhandled brush type: " + brush.type);
  }
};

const applyBrushToGlyphs = (
  glyphs: string[],
  brush: BrushNode,
  context: BuildContext
) => {
  // TODO: Expand to have both GlyphTarget and ScopeTarget, logic for setting
  // scope will be a bit different
  // TODO: Could optimise a little by creating
  // an index up fromnt
  const targets: GlyphTarget[] = [];
  glyphs.forEach((glyph, index) => {
    // TODO: We will need to use VirtualGrid in the end
    // because we might need to set glyphs outside of
    // the initial range
    context.rows.forEach((row, y) => {
      row.forEach((cell, x) => {
        if (cell === glyph) {
          targets.push({ x, y, glyph, index }); // TODO: Is glyph needed?
        }
      });
    });
  });

  applyBrushToGlyphTargets(targets, brush, context);
};

const executeBrushOpNode = (op: BrushOpNode, context: BuildContext) => {
  if (!op.target) {
    throw new Error("Missing target: " + stringify(op, null, "  "));
  }
  switch (op.target.type) {
    case "Heromap::WordNode":
      // storeBrushInScope(op.target.path, op.brush, context);
      break;
    default: {
      let glyphs;
      switch (op.target.type) {
        case "Heromap::GlyphNode":
          glyphs = [op.target.glyph];
          break;
        case "Heromap::GlyphsNode":
          glyphs = op.target.glyphs;
          break;
        default:
          throw new Error("Unhandled op target type " + op.type);
      }
      applyBrushToGlyphs(glyphs, op.brush, context);
    }
  }
};

const executeOp = (op: OperationNode, context: BuildContext) => {
  if (!op || !op.type) {
    throw new Error("Invalid OperationNode: " + stringify(op, null, "  "));
  }
  switch (op.type) {
    case "Heromap::BrushOpNode":
      executeBrushOpNode(op as BrushOpNode, context);
      break;
    default:
      throw new Error("Unhandled node type " + op.type);
  }
};

export type MapElement = {
  type: string;
  params: Record<string, any>;
};

export const build = (
  map: MapNode,
  rng: RNG = createRng(),
  scope: Record<string, any> = {}
) => {
  const width = Math.max(...map.lines.map((line) => line.length));
  const height = map.lines.length;
  const rows = map.lines.map((line) => line.split(""));
  const grid = virtualGrid<MapElement>(width, height);
  const context: BuildContext = {
    map,
    grid,
    rows,
    scope,
    rng,
  };
  for (const op of map.legend) {
    executeOp(op, context);
  }
  // TODO: Any other metadata?
  return {
    // TODO: These bounds haven't been computed
    bounds: {
      x: 0,
      y: 0,
      width: grid.width,
      height: grid.height,
      rows: context.rows,
    },
    cells: grid.getCells(),
  };
};
