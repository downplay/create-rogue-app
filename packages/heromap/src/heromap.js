// Generated automatically by nearley, version 2.20.1
// http://github.com/Hardmath123/nearley
(function () {
function id(x) { return x[0]; }


const moo = require("moo");
const newline = { match: /\n/, lineBreaks: true };
const space = { match: /[^\S\n]+/, lineBreaks: false };

const lexer = moo.states({
    // Consume all whitespace until we see first line that contains at least one non-whitespace char
    begin: {
        newline,
        mapstart: { match: /(?=[\s]*[^\s\n])/, next: "map" },
        space
    },
    map: {
        mapend: { match: /(?=\n\n)/, lineBreaks: true, next: "legend" },
        newline,
        glyphs: { match: /.+/ }
    },
    legend: {
        space,
        "=": "=",
        "[": { match: /\[(?=.)/, push: "glyphs" },
        "|": "|",
        "(": "(",
        ")": ")",
        "+": "+",
        ":": ":",
        "%": "%",
        fraction: { match: /[0-9]+\/[0-9]+/ },
        number: { match: /[0-9]+/ },
        word: { match: /[a-zA-Z][a-zA-Z0-9]+/ },
        glyph: { match: /\S/ },
        newline
    },
    glyphs: {
        escape: { match: "\]", value: x => x[1] },
        "]": { match: "]", pop: 1 },
        glyph: { match: /\S/ },
    }
})

const empty = () => null;

const vid = d => d[0].value;

const heromapSymbol = Symbol("heromap");

const mapNode = ({ map, legend }) => ({
    type: "Heromap::MapNode",
    map,
    legend
})

const brushOperationNode = ({ glyph, brushes }) => ({
    type: "Heromap::BrushOperationNode",
    glyph,
    brushes
})

const operationGroupNode = (operations, quanitifer) => ({
    type: "Heromap::GroupOperationNode",
    operations,
    quanitifer
})

const operationSwitchNode = (switches) => ({
    type: "Heromap::OperationSwitchNode",
    switches
})

const brushesNode = (brushes) => ({
    type: "Heromap::GroupBrushNode",
    brushes
})

const fractionValue = value => {
    const parts = value.split("/");
    return ({ type: "Heromap::FractionValue", numerator: Number(parts[0]), denominator: Number(parts[1]) });
}

const integerValue = value => {
    return ({ type: "Heromap::IntegerValue", value });
}

const percentageValue = value => {
    return ({ type: "Heromap::PercentageValue", value });
}

var grammar = {
    Lexer: lexer,
    ParserRules: [
    {"name": "main", "symbols": ["_", "map", "legend", "_"], "postprocess": d => mapNode({ map: d[1], legend: d[2] })},
    {"name": "map$ebnf$1", "symbols": ["mapline"]},
    {"name": "map$ebnf$1", "symbols": ["map$ebnf$1", "mapline"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "map", "symbols": [(lexer.has("mapstart") ? {type: "mapstart"} : mapstart), "map$ebnf$1"], "postprocess": ([_, mapline]) => mapline},
    {"name": "mapline$ebnf$1", "symbols": [(lexer.has("mapend") ? {type: "mapend"} : mapend)], "postprocess": id},
    {"name": "mapline$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "mapline", "symbols": [(lexer.has("glyphs") ? {type: "glyphs"} : glyphs), "mapline$ebnf$1", (lexer.has("newline") ? {type: "newline"} : newline)], "postprocess": d => d[0].value.trimEnd()},
    {"name": "legend$ebnf$1", "symbols": ["op"]},
    {"name": "legend$ebnf$1", "symbols": ["legend$ebnf$1", "op"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "legend", "symbols": ["legend$ebnf$1"], "postprocess": id},
    {"name": "op", "symbols": ["applyOp"], "postprocess": id},
    {"name": "op", "symbols": ["groupOps"], "postprocess": d => operationSwitchNode(d[0])},
    {"name": "applyOp", "symbols": ["__", "stuff", "__", {"literal":"="}, "__", "brushes"], "postprocess": d => brushOperationNode({glyph: d[1], brushes: brushesNode(d[5]) })},
    {"name": "groupOps", "symbols": ["groupOp"], "postprocess": d => [d[0]]},
    {"name": "groupOps", "symbols": ["groupOps", "__", {"literal":"|"}, "groupOp"], "postprocess": d => [...d[0], d[2]]},
    {"name": "groupOp$ebnf$1", "symbols": ["quantifier"], "postprocess": id},
    {"name": "groupOp$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "groupOp", "symbols": ["__", {"literal":"("}, "legend", "_", {"literal":")"}, "groupOp$ebnf$1"], "postprocess": d => operationGroupNode(d[1], d[6])},
    {"name": "brushes", "symbols": ["brush"], "postprocess": d => [d[0]]},
    {"name": "brushes", "symbols": ["brushes", "__", {"literal":"+"}, "__", "brush"], "postprocess": d => [...d[0], d[4]]},
    {"name": "brush", "symbols": ["switch"], "postprocess": d => ({ switch: d[0] })},
    {"name": "brush", "symbols": ["group"], "postprocess": ([group]) => ({ group })},
    {"name": "switch", "symbols": ["option"], "postprocess": d => [d[0]]},
    {"name": "switch", "symbols": ["switch", "__", {"literal":"|"}, "__", "option"], "postprocess": d => [...d[0], d[4]]},
    {"name": "option$subexpression$1", "symbols": ["stuff"]},
    {"name": "option$subexpression$1", "symbols": ["group"]},
    {"name": "option$ebnf$1", "symbols": ["quantifier"], "postprocess": id},
    {"name": "option$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "option", "symbols": ["option$subexpression$1", "option$ebnf$1"], "postprocess": ([option, quantifier]) => ({ option, quantifier })},
    {"name": "quantifier", "symbols": [{"literal":":"}, "numeric"], "postprocess": d => d[1]},
    {"name": "numeric", "symbols": [(lexer.has("number") ? {type: "number"} : number)], "postprocess": d => integerValue(Number(d[0].value))},
    {"name": "numeric", "symbols": [(lexer.has("fraction") ? {type: "fraction"} : fraction)], "postprocess": d => fractionValue(d[0].value)},
    {"name": "numeric", "symbols": ["percentage"], "postprocess": id},
    {"name": "percentage", "symbols": [(lexer.has("number") ? {type: "number"} : number), {"literal":"%"}], "postprocess": d => percentageValue(Number(d[0]))},
    {"name": "group", "symbols": [{"literal":"("}, "_", "brushes", "_", {"literal":")"}], "postprocess": d => brushesNode(d[2])},
    {"name": "stuff", "symbols": ["glyphs"], "postprocess": d => ({ glyphs: d[0] })},
    {"name": "stuff", "symbols": [(lexer.has("word") ? {type: "word"} : word)], "postprocess": d => ({ word: d[0].value })},
    {"name": "stuff", "symbols": ["glyph"], "postprocess": d => ({ glyph: d[0] })},
    {"name": "glyphs$ebnf$1", "symbols": ["glyph"]},
    {"name": "glyphs$ebnf$1", "symbols": ["glyphs$ebnf$1", "glyph"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "glyphs", "symbols": [{"literal":"["}, "glyphs$ebnf$1", {"literal":"]"}], "postprocess": d => d[1].join()},
    {"name": "glyph", "symbols": [(lexer.has("glyph") ? {type: "glyph"} : glyph)], "postprocess": vid},
    {"name": "glyph", "symbols": [(lexer.has("escape") ? {type: "escape"} : escape)], "postprocess": vid},
    {"name": "glyph", "symbols": [{"literal":"="}], "postprocess": vid},
    {"name": "glyph", "symbols": [{"literal":"|"}], "postprocess": vid},
    {"name": "glyph", "symbols": [{"literal":"+"}], "postprocess": vid},
    {"name": "glyph", "symbols": [{"literal":"["}], "postprocess": vid},
    {"name": "glyph", "symbols": [{"literal":"("}], "postprocess": vid},
    {"name": "glyph", "symbols": [{"literal":")"}], "postprocess": vid},
    {"name": "glyph", "symbols": [{"literal":":"}], "postprocess": vid},
    {"name": "glyph", "symbols": [{"literal":"%"}], "postprocess": vid},
    {"name": "glyph", "symbols": [(lexer.has("number") ? {type: "number"} : number)], "postprocess": (d, location, reject) => d[0].value.length === 1 ? d[0].value.length : reject},
    {"name": "_$ebnf$1", "symbols": []},
    {"name": "_$ebnf$1", "symbols": ["_$ebnf$1", "whitespace"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "_", "symbols": ["_$ebnf$1"], "postprocess": empty},
    {"name": "__$ebnf$1", "symbols": ["whitespace"]},
    {"name": "__$ebnf$1", "symbols": ["__$ebnf$1", "whitespace"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "__", "symbols": ["__$ebnf$1"], "postprocess": empty},
    {"name": "___$ebnf$1", "symbols": []},
    {"name": "___$ebnf$1", "symbols": ["___$ebnf$1", "whitespace"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "___", "symbols": ["___$ebnf$1", (lexer.has("newline") ? {type: "newline"} : newline)], "postprocess": empty},
    {"name": "whitespace", "symbols": [(lexer.has("space") ? {type: "space"} : space)]},
    {"name": "whitespace", "symbols": [(lexer.has("newline") ? {type: "newline"} : newline)], "postprocess": empty}
]
  , ParserStart: "main"
}
if (typeof module !== 'undefined'&& typeof module.exports !== 'undefined') {
   module.exports = grammar;
} else {
   window.grammar = grammar;
}
})();
