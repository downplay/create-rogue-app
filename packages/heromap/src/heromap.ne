@{%

const moo = require("moo");
const newline = { match: /\n/, lineBreaks: true };
const space = { match: /[^\S\n]+/, lineBreaks: false };

const lexer = moo.states({
    // Consume all whitespace until we see first line that contains at least one non-whitespace char
    begin: {
        newline,
        mapstart: { match: /(?=[\s]*[^\s\n])/, next: "map" },
        space
    },
    map: {
        mapend: { match: /(?=\n\n)/, lineBreaks: true, next: "legend" },
        newline,
        glyphs: { match: /.+/ }
    },
    legend: {
        space,
        "=": "=",
        "[": { match: /\[(?=.)/, push: "glyphs" },
        "|": "|",
        "(": "(",
        ")": ")",
        "+": "+",
        ":": ":",
        "%": "%",
        fraction: { match: /[0-9]+\/[0-9]+/ },
        number: { match: /[0-9]+/ },
        word: { match: /[a-zA-Z][a-zA-Z0-9]+/ },
        glyph: { match: /\S/ },
        newline
    },
    glyphs: {
        escape: { match: "\]", value: x => x[1] },
        "]": { match: "]", pop: 1 },
        glyph: { match: /\S/ },
    }
})

const empty = () => null;

const vid = d => d[0].value;

const heromapSymbol = Symbol("heromap");

// TODO: With overlays, make sure trim keeps them all relative
const trimMapLines = (lines) => {
    const min = Math.min(...lines.map(line => line.lastIndexOf(" ")));
    if (min === -1) {
        return lines;
    }
    return lines.map(line => line.slice(min + 1));
}

const mapNode = ({ lines, legend }) => ({
    type: "Heromap::MapNode",
    lines: trimMapLines(lines),
    legend,
    externals: []
})

const brushOpNode = ({ target, brush, op = "apply" }) => ({
    type: "Heromap::BrushOpNode",
    target,
    brush,
    op
})

const operationGroupNode = (operations, quantifier) => ({
    type: "Heromap::AndOpsNode",
    operations,
    quantifier
})

const operationSwitchNode = (switches) => ({
    type: "Heromap::OrOpsNode",
    switches
})

const brushNode = (node, quantifier) => ({
    type: "Heromap::BrushNode",
    brush: node,
    quantifier
})

const andBrushesNode = (brushes) => ({
    type: "Heromap::AndBrushesNode",
    brushes
})

const orBrushesNode = (brushes) => ({
    type: "Heromap::OrBrushesNode",
    brushes
});

const integerValue = value => ({ type: "Heromap::IntegerValue", value });

const fractionValue = value => {
    const parts = value.split("/");
    const denominator = Number(parts[1]);
    if (denominator <= 0) {
        throw new Error("Denominator must be greater than 0");
    }
    return ({ type: "Heromap::FractionValue", numerator: Number(parts[0]), denominator });
}

const decimalValue = value => ({ type: "Heromap::DecimalValue", value });
const percentageValue = value =>  ({ type: "Heromap::PercentageValue", value });

const glyphNode = glyph =>  ({ type: "Heromap::GlyphNode", glyph });
const glyphsNode = glyphs =>  ({ type: "Heromap::GlyphsNode", glyphs: glyphs.split("") });
const wordNode = word =>  ({ type: "Heromap::WordNode", path: [ word ] });

%}

@lexer lexer

main        -> _ lines legend _                           {% d => mapNode({ lines: d[1], legend: d[2] }) %}

lines       -> %mapstart mapline:+                        {% ([_, mapline]) => mapline %}

mapline     -> %glyphs %mapend:? %newline                 {% d => d[0].value.trimEnd() %}

legend      -> op:+                                       {% id %}

op          -> applyOp                                    {% id %}
             | groupOps                                   {% d => operationSwitchNode(d[0]) %}

applyOp     -> __ stuff __ "=" __ brushes            
               {% d => brushOpNode({target: d[1], brush: brushNode(andBrushesNode(d[5])) }) %} 

groupOps    -> groupOp                                    {% d => [d[0]] %}
             | groupOps __ "|" groupOp                    {% d => [...d[0], d[3]] %}

# TODO: COULD make spaces inside the brackets optional here, with a special version of legend/op/applyOp
groupOp     -> __ "(" legend _ ")" quantifier:?           {% d => operationGroupNode(d[2], d[5]) %}

brushes     -> brush                                      {% d => [brushNode(d[0])] %}
             | brushes __ "+" __ brush                    {% d => [...d[0], brushNode(d[4])] %}

brush       -> switch                                     {% d => orBrushesNode(d[0]) %} 
             | group                                      {% id %}

switch      -> option                                     {% d => [d[0]] %}
             | switch __ "|" __ option                    {% d => [...d[0], d[4]] %}

option      -> (stuff | group) quantifier:?               {% d => brushNode(d[0][0], d[1]) %}

quantifier  -> ":" numeric                                {% d => d[1] %}

numeric     -> %number                                    {% d => integerValue(Number(d[0].value)) %}
             | %fraction                                  {% d => fractionValue(d[0].value) %} 
             | percentage                                 {% id %}

# TODO: decimals too
percentage  -> %number "%"                                {% d => percentageValue(Number(d[0])/100) %}

group       -> "(" _ brushes _ ")"                        {% d => andBrushesNode(d[2]) %}

stuff       -> glyphs                                     {% d => glyphsNode(d[0]) %}
             | %word                                      {% d => wordNode(d[0].value) %}
             | glyph                                      {% d => glyphNode(d[0]) %}

# Technically we could allow spaces in a glyph group; probably not wise tho

glyphs      -> "[" glyph:+ "]"                            {% d => d[1].join() %}

glyph       -> %glyph                                     {% vid %}
             | %escape                                    {% vid %}
             | "="                                        {% vid %}
             | "|"                                        {% vid %}
             | "+"                                        {% vid %}
             | "["                                        {% vid %}
             | "("                                        {% vid %}
             | ")"                                        {% vid %}
             | ":"                                        {% vid %}
             | "%"                                        {% vid %}                  
             | %number                                    
               {% (d, location, reject) => d[0].value.length === 1 ? d[0].value.length : reject %} 

_           -> whitespace:*                               {% empty %}

__          -> whitespace:+                               {% empty %}

___         -> whitespace:* %newline                      {% empty %}

whitespace  -> %space | %newline                          {% empty %} 