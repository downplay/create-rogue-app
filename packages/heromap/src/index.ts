export { virtualGrid } from "./virtualGrid"
export type { VirtualGrid } from "./virtualGrid"
export { map, parse } from "./parse"
export { build } from "./build"
export type { MapNode } from "./types"
