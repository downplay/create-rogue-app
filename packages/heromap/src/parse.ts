import * as nearley from "nearley";
import grammar from "./heromap";
import { MapNode } from "./types";

export const parse = (input: string): MapNode => {
  const parser = new nearley.Parser(nearley.Grammar.fromCompiled(grammar));
  let parsed;
  try {
    parsed = parser.feed(input.trim() + "\n");
  } catch (error) {
    console.error("Unparseable text:");
    console.error(input);
    console.error(error);
    throw new Error("Unparseable text");
  }
  if (!parsed.results[0]) {
    throw new Error("Parsed empty map"); // Shouldn't happen
  }
  const map: MapNode = (parsed.results[0] as unknown) as MapNode;
  // Perform some validation?
  return map;
};

let externalIndex = 0;

const INLINE_EXTERNAL_TYPES = ["string", "boolean", "number"];

export const map = (
  input: TemplateStringsArray,
  ...interpolations: any[] // TODO: More strict enforcement of interp types
): MapNode => {
  const externals: Record<string, any> = {};
  const flattened = input
    .map((fragment, i) => {
      if (interpolations[i]) {
        const external = interpolations[i];
        if (external === null || typeof external === "undefined") {
          return fragment;
        }
        if (INLINE_EXTERNAL_TYPES.includes(typeof external)) {
          // TODO: Should boolean true/false be parsed? Or convert to 1/0?
          return fragment + external;
        }
        // TODO: Maybe special handling for some Herotext / Heromap nodes
        // e.g. import maps to render sub-maps
        // TODO: Vector types for x,y placement

        // TODO: Track externals with symbol prop
        const labelName = "OUT" + externalIndex;
        externalIndex++;
        externals[labelName] = external;
        return fragment + labelName;
      }
      return fragment;
    })
    .join("");
  try {
    const main = parse(flattened);
    return {
      ...main,
      externals,
    };
  } catch (e) {
    return errorMain(e.message);
  }
};
