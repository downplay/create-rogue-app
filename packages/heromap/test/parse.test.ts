import { map } from "../src/parse";

it("parses a map", () => {
  const parsed = map`
    ####
    #..#
    ####

    Foo  = Bar
    a    = Baz
    [bc] = [ab]
    b    = Sheep | Cow:30% + Rat:2 | (Bat + Cat):2/3
    
    ( a = b ) | ( Cow = Rat )
    `;

  expect(parsed).toMatchSnapshot();
});
