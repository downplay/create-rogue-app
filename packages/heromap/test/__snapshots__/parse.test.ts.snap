// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses a map 1`] = `
Object {
  "externals": Object {},
  "legend": Array [
    Object {
      "brush": Object {
        "brush": Object {
          "brushes": Array [
            Object {
              "brush": Object {
                "brushes": Array [
                  Object {
                    "brush": Object {
                      "path": Array [
                        "Bar",
                      ],
                      "type": "Heromap::WordNode",
                    },
                    "quantifier": null,
                    "type": "Heromap::BrushNode",
                  },
                ],
                "type": "Heromap::OrBrushesNode",
              },
              "quantifier": undefined,
              "type": "Heromap::BrushNode",
            },
          ],
          "type": "Heromap::AndBrushesNode",
        },
        "quantifier": undefined,
        "type": "Heromap::BrushNode",
      },
      "op": "apply",
      "target": Object {
        "path": Array [
          "Foo",
        ],
        "type": "Heromap::WordNode",
      },
      "type": "Heromap::BrushOpNode",
    },
    Object {
      "brush": Object {
        "brush": Object {
          "brushes": Array [
            Object {
              "brush": Object {
                "brushes": Array [
                  Object {
                    "brush": Object {
                      "path": Array [
                        "Baz",
                      ],
                      "type": "Heromap::WordNode",
                    },
                    "quantifier": null,
                    "type": "Heromap::BrushNode",
                  },
                ],
                "type": "Heromap::OrBrushesNode",
              },
              "quantifier": undefined,
              "type": "Heromap::BrushNode",
            },
          ],
          "type": "Heromap::AndBrushesNode",
        },
        "quantifier": undefined,
        "type": "Heromap::BrushNode",
      },
      "op": "apply",
      "target": Object {
        "glyph": "a",
        "type": "Heromap::GlyphNode",
      },
      "type": "Heromap::BrushOpNode",
    },
    Object {
      "brush": Object {
        "brush": Object {
          "brushes": Array [
            Object {
              "brush": Object {
                "brushes": Array [
                  Object {
                    "brush": Object {
                      "glyphs": Array [
                        "a",
                        ",",
                        "b",
                      ],
                      "type": "Heromap::GlyphsNode",
                    },
                    "quantifier": null,
                    "type": "Heromap::BrushNode",
                  },
                ],
                "type": "Heromap::OrBrushesNode",
              },
              "quantifier": undefined,
              "type": "Heromap::BrushNode",
            },
          ],
          "type": "Heromap::AndBrushesNode",
        },
        "quantifier": undefined,
        "type": "Heromap::BrushNode",
      },
      "op": "apply",
      "target": Object {
        "glyphs": Array [
          "b",
          ",",
          "c",
        ],
        "type": "Heromap::GlyphsNode",
      },
      "type": "Heromap::BrushOpNode",
    },
    Object {
      "brush": Object {
        "brush": Object {
          "brushes": Array [
            Object {
              "brush": Object {
                "brushes": Array [
                  Object {
                    "brush": Object {
                      "path": Array [
                        "Sheep",
                      ],
                      "type": "Heromap::WordNode",
                    },
                    "quantifier": null,
                    "type": "Heromap::BrushNode",
                  },
                  Object {
                    "brush": Object {
                      "path": Array [
                        "Cow",
                      ],
                      "type": "Heromap::WordNode",
                    },
                    "quantifier": Object {
                      "type": "Heromap::PercentageValue",
                      "value": 0.3,
                    },
                    "type": "Heromap::BrushNode",
                  },
                ],
                "type": "Heromap::OrBrushesNode",
              },
              "quantifier": undefined,
              "type": "Heromap::BrushNode",
            },
            Object {
              "brush": Object {
                "brushes": Array [
                  Object {
                    "brush": Object {
                      "path": Array [
                        "Rat",
                      ],
                      "type": "Heromap::WordNode",
                    },
                    "quantifier": Object {
                      "type": "Heromap::IntegerValue",
                      "value": 2,
                    },
                    "type": "Heromap::BrushNode",
                  },
                  Object {
                    "brush": Object {
                      "brushes": Array [
                        Object {
                          "brush": Object {
                            "brushes": Array [
                              Object {
                                "brush": Object {
                                  "path": Array [
                                    "Bat",
                                  ],
                                  "type": "Heromap::WordNode",
                                },
                                "quantifier": null,
                                "type": "Heromap::BrushNode",
                              },
                            ],
                            "type": "Heromap::OrBrushesNode",
                          },
                          "quantifier": undefined,
                          "type": "Heromap::BrushNode",
                        },
                        Object {
                          "brush": Object {
                            "brushes": Array [
                              Object {
                                "brush": Object {
                                  "path": Array [
                                    "Cat",
                                  ],
                                  "type": "Heromap::WordNode",
                                },
                                "quantifier": null,
                                "type": "Heromap::BrushNode",
                              },
                            ],
                            "type": "Heromap::OrBrushesNode",
                          },
                          "quantifier": undefined,
                          "type": "Heromap::BrushNode",
                        },
                      ],
                      "type": "Heromap::AndBrushesNode",
                    },
                    "quantifier": Object {
                      "denominator": 3,
                      "numerator": 2,
                      "type": "Heromap::FractionValue",
                    },
                    "type": "Heromap::BrushNode",
                  },
                ],
                "type": "Heromap::OrBrushesNode",
              },
              "quantifier": undefined,
              "type": "Heromap::BrushNode",
            },
          ],
          "type": "Heromap::AndBrushesNode",
        },
        "quantifier": undefined,
        "type": "Heromap::BrushNode",
      },
      "op": "apply",
      "target": Object {
        "glyph": "b",
        "type": "Heromap::GlyphNode",
      },
      "type": "Heromap::BrushOpNode",
    },
    Object {
      "switches": Array [
        Object {
          "operations": Array [
            Object {
              "brush": Object {
                "brush": Object {
                  "brushes": Array [
                    Object {
                      "brush": Object {
                        "brushes": Array [
                          Object {
                            "brush": Object {
                              "glyph": "b",
                              "type": "Heromap::GlyphNode",
                            },
                            "quantifier": null,
                            "type": "Heromap::BrushNode",
                          },
                        ],
                        "type": "Heromap::OrBrushesNode",
                      },
                      "quantifier": undefined,
                      "type": "Heromap::BrushNode",
                    },
                  ],
                  "type": "Heromap::AndBrushesNode",
                },
                "quantifier": undefined,
                "type": "Heromap::BrushNode",
              },
              "op": "apply",
              "target": Object {
                "glyph": "a",
                "type": "Heromap::GlyphNode",
              },
              "type": "Heromap::BrushOpNode",
            },
          ],
          "quantifier": null,
          "type": "Heromap::AndOpsNode",
        },
        Object {
          "operations": Array [
            Object {
              "brush": Object {
                "brush": Object {
                  "brushes": Array [
                    Object {
                      "brush": Object {
                        "brushes": Array [
                          Object {
                            "brush": Object {
                              "path": Array [
                                "Rat",
                              ],
                              "type": "Heromap::WordNode",
                            },
                            "quantifier": null,
                            "type": "Heromap::BrushNode",
                          },
                        ],
                        "type": "Heromap::OrBrushesNode",
                      },
                      "quantifier": undefined,
                      "type": "Heromap::BrushNode",
                    },
                  ],
                  "type": "Heromap::AndBrushesNode",
                },
                "quantifier": undefined,
                "type": "Heromap::BrushNode",
              },
              "op": "apply",
              "target": Object {
                "path": Array [
                  "Cow",
                ],
                "type": "Heromap::WordNode",
              },
              "type": "Heromap::BrushOpNode",
            },
          ],
          "quantifier": null,
          "type": "Heromap::AndOpsNode",
        },
      ],
      "type": "Heromap::OrOpsNode",
    },
  ],
  "lines": Array [
    "####",
    "#..#",
    "####",
  ],
  "type": "Heromap::MapNode",
}
`;
